python/src/iceberg/catalog/base.py

```diff
@@ -0,0 +1,198 @@
+
+        :param from_name: Existing table's name. Fully classified table name, if it is a namespaced catalog.
+        :param to_name: New Table name to be assigned. Fully classified table name, if it is a namespaced catalog.
+        :return: Nothing
+        :raises TableNotFoundError: If a table with the name does not exist
+        """
+
+    @abstractmethod
+    def replace_table(
+        self,
+        name: str,
+        schema: Schema,
+        partition_spec: PartitionSpec,
+        *,
+        location: Optional[str] = None,
+        properties: Optional[dict] = None
+    ) -> Table:
+        """
+        Starts a transaction and replaces the table with the provided spec.
+
+        :param name: Table's name. Fully classified table name, if it is a namespaced catalog.
+        :param schema: Table's schema
+        :param partition_spec: A partition spec for the table
+        :param location: a location for the table; Optional Keyword Argument
+        :param properties: a string dictionary of table properties; Optional Keyword Argument
+        :return: the replaced table instance
+        :raises TableNotFoundError: If a table with the name does not exist
+        """
+
+
+class NamespacedCatalog(Catalog):
+    """
+    Base catalog for catalogs that support namespaces.
+    """
+
+    @abstractmethod
+    def create_namespace(self, namespace: str, properties: Optional[dict] = None) -> None:
+        """
+        Create a namespace in the catalog.
+
+        :param namespace: The namespace to be created.
+        :param properties: A string dict of properties for the given namespace
+        :return: Nothing
+        :raises AlreadyExistsError: If a namespace with the name already exists in the namespace
+        """
+
+    @abstractmethod
+    def drop_namespace(self, namespace: str) -> None:
+        """
+        Drop a namespace.
+
+        :param namespace: The namespace to be dropped.
+        :return: Nothing
+        :raises NamespaceNotFoundError: If a namespace with the name does not exist in the namespace
+        :raises NamespaceNotEmptyError: If the namespace is not empty
+        """
+
+    @abstractmethod
+    def list_tables(self, namespace: Optional[str] = None) -> list:
+        """
+        List tables under the given namespace in the catalog. If namespace not provided, will list all tables in the
+        catalog.
+
+        :param namespace: the namespace to search
+        :return: list of table names under this namespace.
+        :raises NamespaceNotFoundError: If no such namespace exist
+        """
+
+    @abstractmethod
+    def list_namespaces(self, namespace: Optional[str] = None) -> list:
+        """
+        List namespaces from the given namespace. If not given, list top-level namespaces from the catalog.
+
+        :param namespace: given namespace
+        :return: a List of namespace names
+        """
+
+    @abstractmethod
+    def get_namespace_metadata(self, namespace: str) -> dict:
```

Comment:
I'm not familiar with namespace metadata, is this general metadata that's persisted in the catalog?

Reply:
It is different from catalog metadata. Some (or maybe all) catalogs maintain metadata at a namespace level as well. For instance, JDBC catalog maintains it's namespace level metadata in the table [`iceberg_namespace_properties`](https://github.com/apache/iceberg/blob/96a46c5caa841777d6fff62a17f09dcc6929603d/core/src/main/java/org/apache/iceberg/jdbc/JdbcUtil.java#L81).